### DISCUSSION TITLE
Make email field case insensitive

### DISCUSSION THREAD

**Discussion by aileksandar** [directus/directus#9000] [Archive]
I recently created this issue https://github.com/directus/directus/issues/8984 where there is a bug when checking if the email is unique, and even though it can be an easy fix it got me thinking why do we even need the `checkUniqueEmails` method. It's just doing an additional check for case sensitivity. 

https://github.com/directus/directus/blob/8b362c038326ef2979d05ac628836d507ed07b39/api/src/services/users.ts#L38-L69

My first reaction was to just force all emails to be lowercase, and you can remove the method. But it's not a standard solution, I checked just now on GitHub, and they allow case-insensitive emails. 

But why not just let the database do the work? You can make the email field unique index case-insensitive. It's different for each database, but I think it's possible. 

For MySQL and SQLite, it's just a matter of collation
https://stackoverflow.com/a/6888308/1421641
https://stackoverflow.com/a/20914992/1421641
And for Postgres it can be done via extension or by adding `lower` when creating unique index
http://shuber.io/case-insensitive-unique-constraints-in-postgres/

It would also later maybe make things easier when you need another field like username to also be unique case-insensitive. This is just an idea, and I'm not sure what's the industry standard here, but I would like to hear your thoughts.

**Comment by rijkvanzanten** [comment]
> But why not just let the database do the work? You can make the email field unique index case-insensitive. It's different for each database, but I think it's possible.

This is the reasoning 👍🏻 I don't think it's easily possible in every vendor in a similar / reliable way. We wouldn't want to adjust the users' collation (we're not doing that anywhere else), and different databases will handle those indexes differently. Overall, we're trying to keep the the Directus codebase as "standardized" as possible (meaning I'd like to prevent "If DB == mysql: X" sort of stuff)

**Reply by aileksandar** [reply]
Yeah, it makes sense. I thought you had some contextual checks and actions based on the database vendor. I just thought the logic could be simplified when I saw the size of the method. 

Thanks for the reply, I guess I'll archive the discussion 'till I figure out better solution.

**Reply by rijkvanzanten** [reply]
> I thought you had some contextual checks and actions based on the database vendor.

A couple do exist, that's correct. I long for the day those can be eradicated as well 👀
